### ISSUE TITLE
Use verified L4 kernel instead of Xen

### ISSUE THREAD

**Issue by GWeck** [QubesOS/qubes-issues#3894]
### Qubes OS version:

Far in the Future

### Affected component(s): 

mainly dom0

---

### General notes:

The security of Qubes critically depends on strong isolation provided by Xen. Bugs in Xen endanger the security of Qubes significantly. Possibly the security kernel of L4 (os.inf.tu-dresden.de/L4/) might be used instead of Xen, if vchan and qrexec could be implemented in L4 without too much effort. As L4 is used in very security critical projects, e.g. a filter gateway connecting NATO secret systems to the outside world (https://www.infodas.de/wp-content/uploads/2016/11/SDoT_6.0i_eng_170530.pdf), it is to be expected that using L4 would significantly reduce the risks posed in Xen by bugs.

I think it would be worth while to contact the Technical University of Dresden on this issue. As far as I know, they are currently looking at Qubes and should be interested in a cooperation with the Qubes team. If interested, please contact Prof. Härtig (haertig@os.inf.tu-dresden.de).

---

**Comment by wom-bat** [comment]
You should probably look at seL4 https://seL4.systems/ if you want a proven secure foundation.

**Comment by marmarek** [comment]
See https://www.qubes-os.org/faq/#what-about-this-othernew-microkernelhypervisor

**Comment by GWeck** [comment]
The L4 development team answered the questionnaire as to which features necessary for porting Qubes to another microkernel als follows:
From: Adam Lackorzynski <adam@os.inf.tu-dresden.de>
Subject: Re: Fwd: [QubesOS/qubes-issues] Use verified L4 kernel instead of Xen (#3894)
Date: 26. September 2018 at 18:33:51 CEST
To: Hermann Härtig <haertig@os.inf.tu-dresden.de>
On Wed Sep 26, 2018 at 17:25:57 +0200, Hermann Haertig wrote:

1. What kinds of containers does it use for isolation? Processes? PV VMs? Fully virtualized VMs (HVMs)? And what underlying h/w technology is used (ring0/3, VT-x)?
Fully virtualized VMs (VT-x, AMD-v). Or PV (ring0/3)

2. Does it require specially written/built applications (e.g. patched Firefox)?
Runs unmodified usermode apps (binaries).

3. Does it require custom drivers, or can it use Linux/Windows ones?
Does not require custom drivers, but can be added for special
functionality.

4. Does it support VT-d, and does it allow for the creation of untrusted driver domains?
Yes, and yes.

5. Does it support S3 sleep?
Yes.

6. Does it work on multiple CPUs/Chipsets?
Yes

7. What are the performance costs, more or less? (e.g. “XYZ prevents concurrent execution of two domains/processes on shared cores of a single processor”, etc.)
Typical performance costs due to virtualization.

8. Other special features? E.g. eliminates cooperative covert channels between VMs?
- Microhypervisor based on capability-system
- Built on OS framework, allows for hypervisor/microkernel applications
- real-time capable
- Supports x86, ARM and MIPS
- ...

Adam
---------------------------
From this, it seems that porting should be possible, even without too much effort.
Supporting the L4 kernel might help in building Qubes-based systems in German government projects.

**Comment by GWeck** [comment]
I just provided the answers to the questionnaire as a comment in the
    github issue
    https://github.com/QubesOS/qubes-issues/issues/3894
    
    The L4 team is quite interested, but - like so many of us, lacks
    decent time.
    _____________________________________________________________________

Dr. Gerhard Weck
Franz-Rüth-Str. 2d
50374 Erftstadt
Tel. 02235/44747
E-Mail: GerhardWeck@onlinehome.de
_____________________________________________________________________

    Am 18.09.2018 um 00:12 schrieb Marek
      Marczykowski-Górecki:
    
    
      See https://www.qubes-os.org/faq/#what-about-this-othernew-microkernelhypervisor

**Comment by cfcs** [comment]
@GWeck 
`if vchan and qrexec could be implemented in L4 without too much effort`

It seems to me that if no one can be bothered to do it, it is actually "too much effort."

You could also argue that Genode or Muen would be better than Xen, or that GNU/Herd would be, or that we should retarget to run in QNX. So far nobody but the Qubes team has actually come up with a practical implementation.

If you want to see HAL or Qubes on other platforms than Xen, you should dedicate the time to implement the required feature set (namely shared memory and a stream/message-based primitive), or find funding or people to do that.

As this is an L4 issue, and not a Qubes issue, I believe this issue here should be closed so it doesn't clutter the issue tracker.

**Comment by andrewdavidwong** [comment]
> As this is an L4 issue, and not a Qubes issue, I believe this issue here should be closed so it doesn't clutter the issue tracker.

Leaving it open for now to give the Qubes devs time to consider and decide what to do with it.

**Comment by DemiMarie** [comment]
seL4 does have a vchan implementation, at least.

**Comment by cfcs** [comment]
`We couldn’t find any code matching 'vchan' in seL4/seL4`
@DemiMarie please post a source.

**Comment by DemiMarie** [comment]
@cfcs <https://github.com/seL4/seL4_libs/blob/c7e4a85edc3048979fbc962ace2dbe657e8d0b3c/libsel4vmm/include/vmm/libvchan.h>

Based on the original QubesOS code, too.

**Comment by sptankard** [comment]
Related (because Genode) #5041

**Comment by GWeck** [comment]
For current developments, see also [Makatea: Qubes-like OS on seL4](https://forum.qubes-os.org/t/makatea-qubes-like-os-on-sel4/10353)

**Comment by DemiMarie** [comment]
@GWeck: https://github.com/seL4/seL4/issues/828 will need to be solved first.

**Comment by GWeck** [comment]
I am getting more and more skeptical about this, especially in the context of the x86 architecture. There are so many variants of the processor structure now, and more are still to come - and possibly each of them with different problems with regard to speculative execution. Building a kernel that is verified to work correctly on each of them may show to be impossible. On the other hand, building a kernel that is correct only for some processor versions may be more or less useless, because that version may be obsolete once verification is completed. Other architectures may be better in this respect, e.g. Itanium, which is not troubled by speculative execution problems - as far as I know -, but the IT world has focused on a mediocre architecture - just in accordance with Akerlof's law.

**Comment by DemiMarie** [comment]
@GWeck The long-term solution is speculative taint tracking, which gets the majority (80+%) of the performance benefits of speculative execution without the security problems.  But only new hardware can support it.

**Comment by GWeck** [comment]
... and there we're again on the dependency on processor variants.

**Comment by 0spinboson** [comment]
OT, but how would Akerlof's law be applicable? The main reason Itanium lost out was that Intel hoped to corner the market and become the only 64-bit CPU supplier (through patents etc), but failed partly because they failed to deliver, but also partly because the market didn't go for it, for both cost, compatibility, and lock-in reasons. And given that most of the speculative execution issues we are encountering today affect Intel, fundamentally because Intel largely stopped investing in R&D once AMD made their Bulldozer mistake, and so they never came up with a new and safer core design until now., it seems to me that overall we should consider ourselves lucky that x86 remained dominant, and that AMD came up with x86_64 at the time.

**Comment by GWeck** [comment]
With Itanium, Intel made the same strategic error that Digital made years before with its Alpha processor (which somewhat was Itanium's predecessor). Building a good processor is not enough: You have to provide a migration strategy to this processor which will push the market to it. Digital probably could have changed the market if they had offered their processor, which at that time was the first 64-bit processor, to the gaming industry where it could have been used even then. Just building something for a small market without any perspective to support the mainstream is likely to fail. And the mainstream now sits on a processor design that has been pushed beyond any good architecture and drowns in its ever-rising complexity.

**Comment by andrewdavidwong** [comment]
Please note that the issue tracker (`qubes-issues`) is not intended to serve as a discussion venue. Instead, we've created a designated [forum](https://forum.qubes-os.org/) for discussion and [support](https://www.qubes-os.org/support/). (By contrast, the issue tracker is more of a technical tool intended to support our developers in their work.) Thank you for your understanding!

**Comment by GWeck** [comment]
It's not so much a general discussion but the question of whether this issue can be solved at all.

While replacing Xen with a verified kernel would surely increase the security of Qubes and possibly allow it to be used to separate classified information, this does not come at a low cost. As @DemiMarie pointed out, such a change would require a lot of work to implement features that are now part of Xen. Regarding the very limited development capacities available, the question arises if the available resources should not be better spent elsewhere.

If, as it seems, verification can only be done for each processor variant separately, there is not much benefit in changing the technical basis of Qubes.

So, what should we do with this issue? Close it, or leave for some unspecified time in the future? Any advice would be appreciated.

**Comment by marmarek** [comment]
> Regarding the very limited development capacities available, the question arises if the available resources should not be better spent elsewhere.

That's my opinion about this topic generally. The primary seL4 target use case (embedded and appliance-like) doesn't align well with Qubes OS. Some/most of it can be solved with enough effort, but that's significant amount and as you said - can be better spent elsewhere.

It's useful to have those limitations collected somewhere (also for people asking "why Qubes doesn't use seL4"), but probably open issue is not a good place for that. Maybe collect it into a FAQ entry and close this one? We can of course re-consider in the future, but it seems it isn't viable option at this time (at least for general purpose Qubes OS).

**Comment by DemiMarie** [comment]
> > Regarding the very limited development capacities available, the question arises if the available resources should not be better spent elsewhere.
> 
> That's my opinion about this topic generally. The primary seL4 target use case (embedded and appliance-like) doesn't align well with Qubes OS. Some/most of it can be solved with enough effort, but that's significant amount and as you said - can be better spent elsewhere.
> 
> It's useful to have those limitations collected somewhere (also for people asking "why Qubes doesn't use seL4"), but probably open issue is not a good place for that. Maybe collect it into a FAQ entry and close this one? We can of course re-consider in the future, but it seems it isn't viable option at this time (at least for general purpose Qubes OS).

That is my thought too.  I will also add that seL4 development is ongoing, and I believe the seL4 developers plan to improve dynamic systems support in the future.  Better for us to spend our resources elsewhere and reconsider if/when the seL4 ecosystem has sufficient dynamic systems support.  The work we have done won’t be wasted, after all.

**Comment by andrewdavidwong** [comment]
Sounds like a "won't do."

If anyone has a new reason for why this should be done, please leave a comment, and we'll be happy to take another look. Thank you.

**Comment by DemiMarie** [comment]
Also adding `S: blocked` to indicate that this is blocked on seL4 ecosystem changes.

**Comment by DemiMarie** [comment]
As it turns out, seL4 is [missing mitigations for many speculative execution vulnerabilities][1] and these mitigations [won’t be added][2].  seL4 is a good choice for embedded systems running on processors without such vulnerabilities, but not for Qubes OS.

[1]: https://github.com/seL4/seL4/issues/1108
[2]: https://github.com/seL4/seL4/issues/1108#issuecomment-1736583402
